File: morse-translator-zig/src/state.zig
// ============================================================================
//  Morse-Pi — Shared state, settings, statistics.
//
//  Central state module. Imported by every other module.
//  Uses a global allocator + mutex for thread-safe mutable state.
// ============================================================================
const std = @import("std");

// ── Allocator ────────────────────────────────────────────────────────────────
var gpa_instance = std.heap.GeneralPurposeAllocator(.{}){};
pub const gpa = gpa_instance.allocator();

// ── Mutex (protects settings, state, stats) ──────────────────────────────────
pub var mu: std.Thread.Mutex = .{};

// ════════════════════════════════════════════════════════════════════════════
//  SETTINGS
// ════════════════════════════════════════════════════════════════════════════

pub const Settings = struct {
    speaker_pin: i32 = 18,
    speaker_pin2: ?i32 = null,
    speaker_gnd_mode: []const u8 = "3v3",
    output_type: []const u8 = "speaker",
    pin_mode: []const u8 = "single",
    data_pin: i32 = 17,
    dot_pin: i32 = 22,
    dash_pin: i32 = 27,
    ground_pin: ?i32 = null,
    grounded_pins: []const i32 = &.{},
    use_external_switch: bool = false,
    dot_freq: i32 = 700,
    dash_freq: i32 = 500,
    volume: f64 = 0.75,
    wpm_target: i32 = 20,
    theme: []const u8 = "dark",
    difficulty: []const u8 = "easy",
    quiz_categories: []const []const u8 = &.{"words"},
    practice_chars: []const u8 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789",
    farnsworth_enabled: bool = false,
    farnsworth_letter_mult: f64 = 2.0,
    farnsworth_word_mult: f64 = 2.0,
    device_name: []const u8 = "Morse Pi",
    kb_enabled: bool = false,
    kb_mode: []const u8 = "letters",
    kb_dot_key: []const u8 = "z",
    kb_dash_key: []const u8 = "x",
};

pub var settings: Settings = .{};
pub var settings_json: ?std.json.Parsed(std.json.Value) = null;

/// Raw JSON bytes of current settings (for serving API & templates).
pub var settings_raw: []u8 = &.{};

pub fn loadSettings() void {
    const data = std.fs.cwd().readFileAlloc(gpa, "settings.json", 1 << 20) catch {
        reserializeSettings();
        return;
    };
    defer gpa.free(data);

    // Parse into Value tree (keeps original JSON shape for unknown keys)
    var parsed = std.json.parseFromSlice(std.json.Value, gpa, data, .{
        .ignore_unknown_fields = true,
    }) catch {
        reserializeSettings();
        return;
    };

    // Extract known fields into the typed struct
    if (parsed.value == .object) {
        const obj = &parsed.value.object;
        if (obj.get("speaker_pin")) |v| if (v == .integer) {
            settings.speaker_pin = @intCast(v.integer);
        };
        if (obj.get("speaker_pin2")) |v| {
            settings.speaker_pin2 = if (v == .integer) @as(?i32, @intCast(v.integer)) else null;
        };
        inline for (.{
            .{ "speaker_gnd_mode", "speaker_gnd_mode" },
            .{ "output_type", "output_type" },
            .{ "pin_mode", "pin_mode" },
            .{ "theme", "theme" },
            .{ "difficulty", "difficulty" },
            .{ "practice_chars", "practice_chars" },
            .{ "device_name", "device_name" },
            .{ "kb_mode", "kb_mode" },
            .{ "kb_dot_key", "kb_dot_key" },
            .{ "kb_dash_key", "kb_dash_key" },
        }) |pair| {
            if (obj.get(pair[0])) |v| if (v == .string) {
                @field(settings, pair[1]) = gpa.dupe(u8, v.string) catch pair[1];
            };
        }
        inline for (.{
            .{ "data_pin", "data_pin" },
            .{ "dot_pin", "dot_pin" },
            .{ "dash_pin", "dash_pin" },
            .{ "dot_freq", "dot_freq" },
            .{ "dash_freq", "dash_freq" },
            .{ "wpm_target", "wpm_target" },
        }) |pair| {
            if (obj.get(pair[0])) |v| if (v == .integer) {
                @field(settings, pair[1]) = @intCast(v.integer);
            };
        }
        if (obj.get("ground_pin")) |v| {
            settings.ground_pin = if (v == .integer) @as(?i32, @intCast(v.integer)) else null;
        };
        if (obj.get("volume")) |v| {
            settings.volume = switch (v) {
                .float => v.float,
                .integer => @floatFromInt(v.integer),
                else => 0.75,
            };
        }
        if (obj.get("farnsworth_letter_mult")) |v| {
            settings.farnsworth_letter_mult = switch (v) {
                .float => v.float,
                .integer => @floatFromInt(v.integer),
                else => 2.0,
            };
        }
        if (obj.get("farnsworth_word_mult")) |v| {
            settings.farnsworth_word_mult = switch (v) {
                .float => v.float,
                .integer => @floatFromInt(v.integer),
                else => 2.0,
            };
        }
        if (obj.get("use_external_switch")) |v| if (v == .bool) {
            settings.use_external_switch = v.bool;
        };
        if (obj.get("farnsworth_enabled")) |v| if (v == .bool) {
            settings.farnsworth_enabled = v.bool;
        };
        if (obj.get("kb_enabled")) |v| if (v == .bool) {
            settings.kb_enabled = v.bool;
        };
        // grounded_pins (array of ints)
        if (obj.get("grounded_pins")) |v| if (v == .array) {
            var list = std.ArrayList(i32).init(gpa);
            for (v.array.items) |item| {
                if (item == .integer) {
                    list.append(@intCast(item.integer)) catch {};
                }
            }
            settings.grounded_pins = list.toOwnedSlice() catch &.{};
        };
    }
    parsed.deinit();
    reserializeSettings();
}

pub fn reserializeSettings() void {
    if (settings_raw.len > 0) gpa.free(settings_raw);
    var buf = std.ArrayList(u8).init(gpa);
    std.json.stringify(settings, .{}, buf.writer()) catch {};
    settings_raw = buf.toOwnedSlice() catch &.{};
}

pub fn saveSettings() void {
    reserializeSettings();
    const file = std.fs.cwd().createFile("settings.json", .{}) catch return;
    defer file.close();
    file.writeAll(settings_raw) catch {};
}

// ════════════════════════════════════════════════════════════════════════════
//  STATISTICS
// ════════════════════════════════════════════════════════════════════════════

pub const Stats = struct {
    total_attempts: i64 = 0,
    correct: i64 = 0,
    streak: i64 = 0,
    best_streak: i64 = 0,
    total_chars: i64 = 0,
};

pub var stats: Stats = .{};

pub fn loadStats() void {
    const data = std.fs.cwd().readFileAlloc(gpa, "stats.json", 1 << 20) catch return;
    defer gpa.free(data);
    const parsed = std.json.parseFromSlice(Stats, gpa, data, .{
        .ignore_unknown_fields = true,
    }) catch return;
    stats = parsed.value;
    parsed.deinit();
}

pub fn saveStats() void {
    const file = std.fs.cwd().createFile("stats.json", .{}) catch return;
    defer file.close();
    std.json.stringify(stats, .{}, file.writer()) catch {};
}

// ════════════════════════════════════════════════════════════════════════════
//  RUNTIME STATE (mutable strings backed by gpa)
// ════════════════════════════════════════════════════════════════════════════

pub var mode: []u8 = &.{};
pub var cheat_sheet: bool = false;
pub var current_phrase: []u8 = &.{};
pub var decode_result: []u8 = &.{};
pub var decode_correct_answer: []u8 = &.{};
pub var speed_phrase: []u8 = &.{};
pub var speed_result: []u8 = &.{};
pub var speed_morse_buffer: []u8 = &.{};
pub var speed_morse_output: []u8 = &.{};
pub var send_output: []u8 = &.{};
pub var encode_output: []u8 = &.{};
pub var encode_input: []u8 = &.{};
pub var button_active: bool = false;
pub var current_morse_buffer: []u8 = &.{};
pub var net_key_mode_enabled: bool = false;
pub var net_morse_buffer_str: []u8 = &.{};
pub var net_morse_output_str: []u8 = &.{};
pub var kb_output: []u8 = &.{};

/// Replace an allocated state string.
pub fn setStr(dest: *[]u8, new_val: []const u8) void {
    if (dest.len > 0) gpa.free(dest.*);
    dest.* = gpa.dupe(u8, new_val) catch &.{};
}

/// Append to an allocated state string.
pub fn appendStr(dest: *[]u8, suffix: []const u8) void {
    if (suffix.len == 0) return;
    const new = gpa.alloc(u8, dest.len + suffix.len) catch return;
    @memcpy(new[0..dest.len], dest.*);
    @memcpy(new[dest.len..], suffix);
    if (dest.len > 0) gpa.free(dest.*);
    dest.* = new;
}

// ── Network inbox ────────────────────────────────────────────────────────
pub const NetMessage = struct {
    sender: []const u8,
    text: []const u8,
    morse: []const u8,
    ts: f64,
};

pub var net_inbox: [32]?NetMessage = .{null} ** 32;
pub var net_inbox_len: usize = 0;

pub fn addInboxMessage(sender: []const u8, text: []const u8, morse_str: []const u8) void {
    const msg = NetMessage{
        .sender = gpa.dupe(u8, sender) catch "?",
        .text = gpa.dupe(u8, text) catch "",
        .morse = gpa.dupe(u8, morse_str) catch "",
        .ts = @as(f64, @floatFromInt(std.time.milliTimestamp())) / 1000.0,
    };
    if (net_inbox_len < 32) {
        net_inbox[net_inbox_len] = msg;
        net_inbox_len += 1;
    } else {
        // shift left, drop oldest
        var i: usize = 0;
        while (i < 31) : (i += 1) net_inbox[i] = net_inbox[i + 1];
        net_inbox[31] = msg;
    }
}

pub fn clearInbox() void {
    net_inbox_len = 0;
    for (&net_inbox) |*slot| slot.* = null;
}

// ── GPIO state ───────────────────────────────────────────────────────────
pub var gpio_available: bool = false;
pub var gpio_error: ?[]const u8 = null;
pub var gpio_error_log: std.ArrayList([]const u8) = std.ArrayList([]const u8).init(gpa);

pub var pin_states: [28]?bool = .{null} ** 28; // indexed by BCM number

// ── pigpio handle ────────────────────────────────────────────────────────
pub var pi_handle: i32 = -1;
pub var wave_id: i32 = -1;
pub var spk_pin1: ?i32 = null;
pub var spk_pin2: ?i32 = null;

// ── Initialisation ───────────────────────────────────────────────────────
pub fn init() void {
    loadSettings();
    loadStats();
    setStr(&mode, "send");
}

// ── JSON helpers ─────────────────────────────────────────────────────────

/// Write full runtime state as JSON (matches Python's shared.state dict).
pub fn writeStateJson(w: anytype) !void {
    try w.writeAll("{");
    try std.fmt.format(w,
        \\"mode":"{s}","cheat_sheet":{s},"current_phrase":"{s}","decode_result":"{s}",
    , .{ mode, if (cheat_sheet) "true" else "false", current_phrase, decode_result });
    try std.fmt.format(w,
        \\"decode_correct_answer":"{s}","speed_phrase":"{s}","speed_result":"{s}",
    , .{ decode_correct_answer, speed_phrase, speed_result });
    try std.fmt.format(w,
        \\"speed_morse_buffer":"{s}","speed_morse_output":"{s}",
    , .{ speed_morse_buffer, speed_morse_output });
    try std.fmt.format(w,
        \\"send_output":"{s}","encode_output":"{s}","encode_input":"{s}",
    , .{ send_output, encode_output, encode_input });
    try std.fmt.format(w,
        \\"button_active":{s},"current_morse_buffer":"{s}",
    , .{ if (button_active) "true" else "false", current_morse_buffer });
    try std.fmt.format(w,
        \\"net_key_mode":{s},"net_morse_buffer":"{s}","net_morse_output":"{s}",
    , .{ if (net_key_mode_enabled) "true" else "false", net_morse_buffer_str, net_morse_output_str });
    try std.fmt.format(w,
        \\"kb_output":"{s}",
    , .{kb_output});

    // stats
    try std.fmt.format(w,
        \\"stats":{{"total_attempts":{},"correct":{},"streak":{},"best_streak":{},"total_chars":{},"sessions":[]}}
    , .{ stats.total_attempts, stats.correct, stats.streak, stats.best_streak, stats.total_chars });

    // net_inbox
    try w.writeAll(",\"net_inbox\":[");
    var first = true;
    for (net_inbox[0..net_inbox_len]) |maybe_msg| {
        if (maybe_msg) |msg| {
            if (!first) try w.writeAll(",");
            first = false;
            try std.fmt.format(w,
                \\{{"sender":"{s}","text":"{s}","morse":"{s}","ts":{d}}}
            , .{ msg.sender, msg.text, msg.morse, msg.ts });
        }
    }
    try w.writeAll("]}");
}

/// Write stats object as JSON.
pub fn writeStatsJson(w: anytype) !void {
    try std.fmt.format(w,
        \\{{"total_attempts":{},"correct":{},"streak":{},"best_streak":{},"total_chars":{},"sessions":[]}}
    , .{ stats.total_attempts, stats.correct, stats.streak, stats.best_streak, stats.total_chars });
}

File: morse-translator-rust/src/state.rs
// ============================================================================
//  Morse-Pi — Shared state, settings, statistics.
//
//  Central state module. Uses Arc<Mutex<...>> for thread-safe mutable state.
// ============================================================================
use serde::{Deserialize, Serialize};
use std::fs;
use std::sync::Mutex;

// ════════════════════════════════════════════════════════════════════════════
//  SETTINGS
// ════════════════════════════════════════════════════════════════════════════

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Settings {
    #[serde(default = "default_speaker_pin")]
    pub speaker_pin: i32,
    #[serde(default)]
    pub speaker_pin2: Option<i32>,
    #[serde(default = "default_speaker_gnd_mode")]
    pub speaker_gnd_mode: String,
    #[serde(default = "default_output_type")]
    pub output_type: String,
    #[serde(default = "default_pin_mode")]
    pub pin_mode: String,
    #[serde(default = "default_data_pin")]
    pub data_pin: i32,
    #[serde(default = "default_dot_pin")]
    pub dot_pin: i32,
    #[serde(default = "default_dash_pin")]
    pub dash_pin: i32,
    #[serde(default)]
    pub ground_pin: Option<i32>,
    #[serde(default)]
    pub grounded_pins: Vec<i32>,
    #[serde(default)]
    pub use_external_switch: bool,
    #[serde(default = "default_dot_freq")]
    pub dot_freq: i32,
    #[serde(default = "default_dash_freq")]
    pub dash_freq: i32,
    #[serde(default = "default_volume")]
    pub volume: f64,
    #[serde(default = "default_wpm_target")]
    pub wpm_target: i32,
    #[serde(default = "default_farnsworth_wpm")]
    pub farnsworth_wpm: i32,
    #[serde(default = "default_theme")]
    pub theme: String,
    #[serde(default = "default_difficulty")]
    pub difficulty: String,
    #[serde(default = "default_quiz_categories")]
    pub quiz_categories: Vec<String>,
    #[serde(default = "default_practice_chars")]
    pub practice_chars: String,
    #[serde(default, skip_serializing)]
    pub farnsworth_enabled: bool,
    #[serde(default = "default_farnsworth_mult", skip_serializing)]
    pub farnsworth_letter_mult: f64,
    #[serde(default = "default_farnsworth_mult", skip_serializing)]
    pub farnsworth_word_mult: f64,
    #[serde(default = "default_device_name")]
    pub device_name: String,
    #[serde(default)]
    pub kb_enabled: bool,
    #[serde(default = "default_kb_mode")]
    pub kb_mode: String,
    #[serde(default = "default_kb_dot_key")]
    pub kb_dot_key: String,
    #[serde(default = "default_kb_dash_key")]
    pub kb_dash_key: String,
}

fn default_speaker_pin() -> i32 { 18 }
fn default_speaker_gnd_mode() -> String { "3v3".into() }
fn default_output_type() -> String { "speaker".into() }
fn default_pin_mode() -> String { "single".into() }
fn default_data_pin() -> i32 { 17 }
fn default_dot_pin() -> i32 { 22 }
fn default_dash_pin() -> i32 { 27 }
fn default_dot_freq() -> i32 { 700 }
fn default_dash_freq() -> i32 { 500 }
fn default_volume() -> f64 { 0.75 }
fn default_wpm_target() -> i32 { 20 }
fn default_farnsworth_wpm() -> i32 { 20 }
fn default_theme() -> String { "dark".into() }
fn default_difficulty() -> String { "easy".into() }
fn default_quiz_categories() -> Vec<String> { vec!["words".into()] }
fn default_practice_chars() -> String { "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".into() }
fn default_farnsworth_mult() -> f64 { 2.0 }
fn default_device_name() -> String { "Morse Pi".into() }
fn default_kb_mode() -> String { "letters".into() }
fn default_kb_dot_key() -> String { "z".into() }
fn default_kb_dash_key() -> String { "x".into() }

impl Default for Settings {
    fn default() -> Self {
        Self {
            speaker_pin: 18,
            speaker_pin2: None,
            speaker_gnd_mode: "3v3".into(),
            output_type: "speaker".into(),
            pin_mode: "single".into(),
            data_pin: 17,
            dot_pin: 22,
            dash_pin: 27,
            ground_pin: None,
            grounded_pins: vec![],
            use_external_switch: false,
            dot_freq: 700,
            dash_freq: 500,
            volume: 0.75,
            wpm_target: 20,
            farnsworth_wpm: 20,
            theme: "dark".into(),
            difficulty: "easy".into(),
            quiz_categories: vec!["words".into()],
            practice_chars: "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".into(),
            farnsworth_enabled: false,
            farnsworth_letter_mult: 2.0,
            farnsworth_word_mult: 2.0,
            device_name: "Morse Pi".into(),
            kb_enabled: false,
            kb_mode: "letters".into(),
            kb_dot_key: "z".into(),
            kb_dash_key: "x".into(),
        }
    }
}

// ════════════════════════════════════════════════════════════════════════════
//  STATISTICS
// ════════════════════════════════════════════════════════════════════════════

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Stats {
    #[serde(default)]
    pub total_attempts: i64,
    #[serde(default)]
    pub correct: i64,
    #[serde(default)]
    pub streak: i64,
    #[serde(default)]
    pub best_streak: i64,
    #[serde(default)]
    pub total_chars: i64,
}

// ════════════════════════════════════════════════════════════════════════════
//  NETWORK MESSAGE
// ════════════════════════════════════════════════════════════════════════════

#[derive(Debug, Clone, Serialize)]
pub struct NetMessage {
    pub sender: String,
    pub text: String,
    pub morse: String,
    pub ts: f64,
}

// ════════════════════════════════════════════════════════════════════════════
//  RUNTIME STATE
// ════════════════════════════════════════════════════════════════════════════

pub struct AppState {
    pub settings: Settings,
    pub stats: Stats,
    pub mode: String,
    pub cheat_sheet: bool,
    pub current_phrase: String,
    pub decode_result: String,
    pub decode_correct_answer: String,
    pub speed_phrase: String,
    pub speed_result: String,
    pub speed_morse_buffer: String,
    pub speed_morse_output: String,
    pub send_output: String,
    pub encode_output: String,
    pub encode_input: String,
    pub button_active: bool,
    pub current_morse_buffer: String,
    pub net_key_mode_enabled: bool,
    pub net_morse_buffer_str: String,
    pub net_morse_output_str: String,
    pub net_live_transmit_enabled: bool,
    pub net_live_transmit_target_ip: String,
    pub net_live_transmit_target_port: u16,
    pub kb_output: String,
    pub net_inbox: Vec<NetMessage>,
    // GPIO state
    pub gpio_available: bool,
    pub gpio_error: Option<String>,
    pub gpio_error_log: Vec<String>,
    pub pin_states: [Option<bool>; 28],
    // pigpio handle
    pub pi_handle: i32,
    pub wave_id: i32,
    pub spk_pin1: Option<i32>,
    pub spk_pin2: Option<i32>,
}

impl Default for AppState {
    fn default() -> Self {
        Self {
            settings: Settings::default(),
            stats: Stats::default(),
            mode: "send".into(),
            cheat_sheet: false,
            current_phrase: String::new(),
            decode_result: String::new(),
            decode_correct_answer: String::new(),
            speed_phrase: String::new(),
            speed_result: String::new(),
            speed_morse_buffer: String::new(),
            speed_morse_output: String::new(),
            send_output: String::new(),
            encode_output: String::new(),
            encode_input: String::new(),
            button_active: false,
            current_morse_buffer: String::new(),
            net_key_mode_enabled: false,
            net_morse_buffer_str: String::new(),
            net_morse_output_str: String::new(),
            net_live_transmit_enabled: false,
            net_live_transmit_target_ip: String::new(),
            net_live_transmit_target_port: 5000,
            kb_output: String::new(),
            net_inbox: Vec::new(),
            gpio_available: false,
            gpio_error: None,
            gpio_error_log: Vec::new(),
            pin_states: [None; 28],
            pi_handle: -1,
            wave_id: -1,
            spk_pin1: None,
            spk_pin2: None,
        }
    }
}

impl AppState {
    pub fn load_settings(&mut self) {
        match fs::read_to_string("settings.json") {
            Ok(data) => match serde_json::from_str::<Settings>(&data) {
                Ok(mut s) => {
                    s.wpm_target = s.wpm_target.max(5);

                    if !data.contains("\"farnsworth_wpm\"") {
                        if s.farnsworth_enabled {
                            let legacy_mult = s
                                .farnsworth_letter_mult
                                .max(s.farnsworth_word_mult)
                                .max(1.0);
                            s.farnsworth_wpm = ((s.wpm_target as f64) / legacy_mult).round() as i32;
                        } else {
                            s.farnsworth_wpm = s.wpm_target;
                        }
                    }

                    s.farnsworth_wpm = s.farnsworth_wpm.max(5).min(s.wpm_target);
                    self.settings = s;
                }
                Err(_) => {}
            },
            Err(_) => {}
        }
    }

    pub fn save_settings(&self) {
        if let Ok(data) = serde_json::to_string_pretty(&self.settings) {
            let _ = fs::write("settings.json", data);
        }
    }

    pub fn settings_json(&self) -> String {
        serde_json::to_string(&self.settings).unwrap_or_else(|_| "{}".into())
    }

    pub fn load_stats(&mut self) {
        match fs::read_to_string("stats.json") {
            Ok(data) => match serde_json::from_str::<Stats>(&data) {
                Ok(s) => self.stats = s,
                Err(_) => {}
            },
            Err(_) => {}
        }
    }

    pub fn save_stats(&self) {
        if let Ok(data) = serde_json::to_string(&self.stats) {
            let _ = fs::write("stats.json", data);
        }
    }

    pub fn add_inbox_message(&mut self, sender: &str, text: &str, morse_str: &str) {
        let msg = NetMessage {
            sender: sender.to_string(),
            text: text.to_string(),
            morse: morse_str.to_string(),
            ts: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|d| d.as_secs_f64())
                .unwrap_or(0.0),
        };
        if self.net_inbox.len() >= 32 {
            self.net_inbox.remove(0);
        }
        self.net_inbox.push(msg);
    }

    pub fn clear_inbox(&mut self) {
        self.net_inbox.clear();
    }

    /// Write full runtime state as JSON.
    pub fn state_json(&self) -> String {
        let stats_json = format!(
            r#"{{"total_attempts":{},"correct":{},"streak":{},"best_streak":{},"total_chars":{},"sessions":[]}}"#,
            self.stats.total_attempts, self.stats.correct, self.stats.streak,
            self.stats.best_streak, self.stats.total_chars,
        );
        let inbox_json = self.inbox_json();
        format!(
            r#"{{"mode":"{}","cheat_sheet":{},"current_phrase":"{}","decode_result":"{}","decode_correct_answer":"{}","speed_phrase":"{}","speed_result":"{}","speed_morse_buffer":"{}","speed_morse_output":"{}","send_output":"{}","encode_output":"{}","encode_input":"{}","button_active":{},"current_morse_buffer":"{}","net_key_mode":{},"net_morse_buffer":"{}","net_morse_output":"{}","net_live_transmit_enabled":{},"kb_output":"{}","stats":{},"net_inbox":{}}}"#,
            escape_json(&self.mode),
            self.cheat_sheet,
            escape_json(&self.current_phrase),
            escape_json(&self.decode_result),
            escape_json(&self.decode_correct_answer),
            escape_json(&self.speed_phrase),
            escape_json(&self.speed_result),
            escape_json(&self.speed_morse_buffer),
            escape_json(&self.speed_morse_output),
            escape_json(&self.send_output),
            escape_json(&self.encode_output),
            escape_json(&self.encode_input),
            self.button_active,
            escape_json(&self.current_morse_buffer),
            self.net_key_mode_enabled,
            escape_json(&self.net_morse_buffer_str),
            escape_json(&self.net_morse_output_str),
            self.net_live_transmit_enabled,
            escape_json(&self.kb_output),
            stats_json,
            inbox_json,
        )
    }

    pub fn stats_json(&self) -> String {
        format!(
            r#"{{"total_attempts":{},"correct":{},"streak":{},"best_streak":{},"total_chars":{},"sessions":[]}}"#,
            self.stats.total_attempts, self.stats.correct, self.stats.streak,
            self.stats.best_streak, self.stats.total_chars,
        )
    }

    fn inbox_json(&self) -> String {
        let entries: Vec<String> = self.net_inbox.iter().map(|msg| {
            format!(
                r#"{{"sender":"{}","text":"{}","morse":"{}","ts":{}}}"#,
                escape_json(&msg.sender),
                escape_json(&msg.text),
                escape_json(&msg.morse),
                msg.ts,
            )
        }).collect();
        format!("[{}]", entries.join(","))
    }

    pub fn init(&mut self) {
        self.load_settings();
        self.load_stats();
        self.mode = "send".into();
    }
}

/// Escape a string for JSON embedding (handle backslash, quotes, newlines).
pub fn escape_json(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for c in s.chars() {
        match c {
            '\\' => out.push_str("\\\\"),
            '"' => out.push_str("\\\""),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            _ => out.push(c),
        }
    }
    out
}

/// Global shared state.
pub static STATE: once_cell::sync::Lazy<Mutex<AppState>> =
    once_cell::sync::Lazy::new(|| Mutex::new(AppState::default()));
